rfcs/0006-configuration-constants.md

```diff
@@ -0,0 +1,99 @@
+
+The configuration variables for a program can be identified by the `configurable` keyword within the Sway language. This denotes a constant within the program that has no known value at compile time. The value will be provided at configuration time by the SDK or some other consumer of the configuration-time-constant descriptor. As an example:
+
+```rust
+// main.sw
+script;
+
+configurable CONTRACT_ADDRESS: b256;
+
+fn main() { ... }
+```
+```
+// configurable-constants.json
+
+{
+    "CONTRACT_ADDRESS": {
+        "type": "b256",
+        "offset": "12345"
+    }
+}
+```
+
+If you are not using the SDK, it will be possible to define these values either by passing a flag to `forc` or including them in the `Forc.toml` manifest file.
+
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The `type` field of the JSON file should describe the type of the configurable in the same manner as types are described in the JSON ABI file.
+
+No two configurables in a program should have the same name to prevent collisions.
+
+The range from the offset to the offset plus the size of the type will represent bytes within the bytecode that will be overwritten by the SDK, which will write values according to the ABI encoder's memory layout, which is consistent with Sway's memory layout.
+
+The change is not breaking as it is entirely new functionality.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+One major drawback is the additional cognitive complexity this adds to the compilation process. If any bug is introduced or the overwriting of the bytecode goes awry in any way, it will result in confusing and inconsistent undefined behavior. We will need to introduce sufficient checks to ensure that an end user would never encounter this situation.
+
+This increases our reliance on the correctness of the ABI encoder in the SDK, and any version mismatches or bugs in the encoder will result in similarly undefined and unpredictable behavior.
+
+This change also further couples the language to the SDK, further minimizing the use case of writing Sway without the SDK. This can be alleviated by introducing mechanisms to `forc` for defining these values manually, as mentioned in the guide-level explanation.
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+One alternative is to introduce an environment variables/configuration-time constants section to the transaction format. This could be cleaner, but the impact on Sway would be minimal. Sway would still output a descriptor file, introduce the `configurable` keyword syntax, etc., and the majority of this RFC would still apply. The benefit of adding this to the transaction format would be potential simplification of the process. A couple negatives would be additional churn on the client, VM, and SDK; and further coupling of specifically Sway to the FuelVM.
```

Comment:
Another alternative would be to get `configurable` values from the VM via a new operation.

Like now (i.e., as in this RFC), the compiler would, in the descriptor file, describe the `configurable`, and in addition provide an integer ID to it. This integer ID would be the argument to the new VM op, and the VM would return the value. The downside to this would be that we have an OP whose return type is not fixed. At the compiler level this is okay because the compiler knows the return type (based on the ID passed) and do its type-checking job.

Reply:
This is a very interesting idea. However, that sounds like runtime environment variables. Here, our objective is to introduce variables after compile time but _before runtime_. The deployed bytecode should not have any variability defined at runtime -- the constants are provided by the SDK during or before the deploy.